[Lens] Revisit flyout fields style from column to row compressed

dej611 opened the pull request:
## Summary

Fixes #118391
Fixes #103778

Mostly changed field aspect from `columnCompressed` to `rowCompressed` with additional refactoring of `EuiSpacer`s, and Top Values Rank direction component.
Updated test to match the new components.

<img width="318" alt="Screenshot 2021-12-01 at 14 25 31" src="https://user-images.githubusercontent.com/924948/144246233-3daae1b5-57ca-41ed-be7b-21abf7782102.png">

<img width="323" alt="Screenshot 2021-12-01 at 14 25 38" src="https://user-images.githubusercontent.com/924948/144246252-e5d9b446-91a1-47cf-91a0-7167db18d938.png">

<img width="320" alt="Screenshot 2021-12-01 at 14 23 38" src="https://user-images.githubusercontent.com/924948/144246273-6eac4894-9420-45de-9350-a5e7907b0250.png">
<img width="318" alt="Screenshot 2021-12-01 at 14 23 49" src="https://user-images.githubusercontent.com/924948/144246276-c25b9773-6c60-41b1-9f14-d91551c5ebfe.png">
<img width="316" alt="Screenshot 2021-12-01 at 14 23 58" src="https://user-images.githubusercontent.com/924948/144246279-2da9fc64-7a84-4010-b384-be51ed0f303d.png">
<img width="321" alt="Screenshot 2021-12-01 at 14 24 08" src="https://user-images.githubusercontent.com/924948/144246280-005a32bd-fbbb-4c17-9de5-05ed344b9f5e.png">
<img width="321" alt="Screenshot 2021-12-01 at 14 24 16" src="https://user-images.githubusercontent.com/924948/144246282-6575e1dd-9b2e-4da2-acaa-07b7085e3eae.png">
<img width="320" alt="Screenshot 2021-12-01 at 14 24 25" src="https://user-images.githubusercontent.com/924948/144246284-0baeea05-708d-4597-b110-73590ed28f98.png">
<img width="317" alt="Screenshot 2021-12-01 at 14 24 33" src="https://user-images.githubusercontent.com/924948/144246286-e88c8f99-35d5-4806-9cc7-7bedadba3e79.png">
<img width="318" alt="Screenshot 2021-12-01 at 14 24 43" src="https://user-images.githubusercontent.com/924948/144246287-be0839e9-a81a-4806-bdca-05898b7c72de.png">
<img width="317" alt="Screenshot 2021-12-01 at 14 24 52" src="https://user-images.githubusercontent.com/924948/144246289-085f3e54-b299-48cc-9f32-1764a7f0ff53.png">
<img width="314" alt="Screenshot 2021-12-01 at 14 25 04" src="https://user-images.githubusercontent.com/924948/144246291-76007170-c584-406c-8b6f-dbee49a3eff9.png">
<img width="317" alt="Screenshot 2021-12-01 at 14 25 16" src="https://user-images.githubusercontent.com/924948/144246292-512664e4-0e16-4f82-863e-2c547e35bb40.png">

## Some questions

* given the increased space for form fields, would it make sense to move some of the numeric fields - when possible - to range sliders?
  * For instance `Top values > Number of values` has both min and max defined. But other fields with min/max defined  (i.e. `Percentile`) can apply as well.

cc @MichaelMarcialis 

### Checklist

Delete any items that are not applicable to this PR.

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [ ] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [ ] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)


### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Multiple Spaces&mdash;unexpected behavior in non-default Kibana Space. | Low | High | Integration tests will verify that all features are still supported in non-default Kibana Space and when user switches between spaces. |
| Multiple nodes&mdash;Elasticsearch polling might have race conditions when multiple Kibana nodes are polling for the same tasks. | High | Low | Tasks are idempotent, so executing them multiple times will not result in logical error, but will degrade performance. To test for this case we add plenty of unit tests around this logic and document manual testing procedure. |
| Code should gracefully handle cases when feature X or plugin Y are disabled. | Medium | High | Unit tests will verify that any feature flag or plugin combination still results in our service operational. |
| [See more potential risk examples](https://github.com/elastic/kibana/blob/main/RISK_MATRIX.mdx) |


### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

elasticmachine commented:
Pinging @elastic/kibana-vis-editors @elastic/kibana-vis-editors-external (Team:VisEditors)

MichaelMarcialis commented:
> * given the increased space for form fields, would it make sense to move some of the numeric fields - when possible - to range sliders?
>   
>   * For instance `Top values > Number of values` has both min and max defined. But other fields with min/max defined  (i.e. `Percentile`) can apply as well.

I'd be OK with adding a supporting range slider to the number inputs assuming 1) there is enough space (i.e. using row compressed) and 2) the related number inputs do not support an infinite range of numbers (i.e. they have a finite min and max). I just want to make sure we're not relying on sliders alone for items that require precision, as the sliders aren't particularly good at that.

mbondyra commented:
This is not caused by your PR, but could you also fix mobile view for the flyout?
<img width="498" alt="Screenshot 2021-12-02 at 11 31 41" src="https://user-images.githubusercontent.com/4283304/144405167-44e535e4-6b13-4337-9564-11f8bc4d6d07.png">
 We could either overwrite the max-width style to 100%: 
<img width="358" alt="Screenshot 2021-12-02 at 11 32 07" src="https://user-images.githubusercontent.com/4283304/144405196-5d1ec1e3-a265-46a3-a499-2b1a398ad18a.png">
or adding to mobile media query `.lnsDimensionContainer` `left: 10vw;` to align it to the right.

dej611 commented:
Fixed the mobile layout for the flyout as proposed by @mbondyra :
<img width="379" alt="Screenshot 2021-12-02 at 12 54 45" src="https://user-images.githubusercontent.com/924948/144417435-4605610c-c358-4614-ace7-766f51bd3ea5.png">

Replaced the Percentile numeric input with a range input:

<img width="325" alt="Screenshot 2021-12-02 at 10 41 47" src="https://user-images.githubusercontent.com/924948/144417503-1e69b783-26e7-417b-9db6-6cf9269eaacb.png">

While the idea of replacing the `Top values > Number of values` sounded nice, when implemented I've realized that:
* it doesn't look that good for the best practices of having low values
* it encourages to set higher values for the function

<img width="324" alt="Screenshot 2021-12-02 at 10 23 20" src="https://user-images.githubusercontent.com/924948/144417658-42f4e668-fa67-489c-8b18-bc39f5ae255c.png">

flash1293 commented:
@dej611 With your additions for mobile does this fix https://github.com/elastic/kibana/issues/103778 as well?

dej611 commented:
> @dej611 With your additions for mobile does this fix #103778 as well?

That looks pretty close @mbondyra 's issue. I'll add the `z-index` thingy to sortout the search bar problem and link the issue in the main description of the PR 🎉

dej611 commented:
Sorted also the z-index problem with the search bar:

<img width="370" alt="Screenshot 2021-12-02 at 15 09 03" src="https://user-images.githubusercontent.com/924948/144438085-2e82821b-f191-472a-99c9-53ec4e093231.png">

dej611 commented:
@elasticmachine merge upstream

mbondyra commented:
Tested on FF and it looks as expected. Approved!

mbondyra commented on x-pack/plugins/lens/public/indexpattern_datasource/dimension_panel/advanced_options.tsx:
```diff
@@ -26,7 +26,7 @@ export function AdvancedOptions(props: {
       {popoverOptions.length > 0 && (
         <EuiText textAlign="right">
-          <EuiSpacer size="s" />
+          {/* <EuiSpacer size="s" /> */}
```
nit: can it be removed?

ryankeairns commented on x-pack/plugins/lens/public/editor_frame_service/editor_frame/config_panel/dimension_container.scss:
```diff
@@ -16,6 +16,8 @@
 
   @include euiBreakpoint('xs', 's', 'm') {
     @include euiFlyout;
+    left: 10vw;
```
This works, but it seems to be overriding some default EUI behavior that should have this flyout working in the desired manner. The issue seems to be the `left: 0` on line 9 above. My guess is that line 9 seeks to ensure the flyout covers the full side panel as opposed to setting a fixed width/large flyout size. 

@MichaelMarcialis may have some better ideas, but it seems moving the `left: 0` into the `'l', 'xl'` breakbpoint will allow the flyout to honor the default EUI stylings for small screens.

MichaelMarcialis replied:
@ryankeairns' analysis is correct. At the same time, I'm not super keen on showing the rest of the Lens interface in the background when the configuration flyouts are open on smaller viewport sizes (looks a bit overwhelming and messy). To address both, I've put together a quick [design PR](https://github.com/dej611/kibana/pull/14) for you @dej611. Have a look when you get a moment, and feel free to merge in.

MichaelMarcialis reviewed (approved):
Great work on this, @dej611. As mentioned in my comment below, I made a quick [design PR](https://github.com/dej611/kibana/pull/14) for the flyout issue on smaller viewport sizes. Once that PR is merged into this branch, that will address my only concerns. Otherwise, this looks good to go. Approving now, assuming you'll be able to merge that design PR in without issue.

mbondyra commented:
@elasticmachine merge upstream
